rfcs/0000_project_typing_guidelines.md

```diff
@@ -0,0 +1,164 @@
+def double_list(l: List[int]) -> List[int]
+    return [x * 2 for x in l]
+
+# Instead, use a collection type
+def double_coll(c: Collection[int]) -> List[int]
+    return [x * 2 for x in c]
+```
+
+Using the broadest type that provides the functionality needed allows for more flexibility when calling the function or method. The second example can be called with a set, while the first only works for lists.
+
+### Methods and functions SHOULD return the narrowest possible type
+
+The type returned from a method or function should be the most concrete type possible. Similar to the previous example, the return type could be `Collection[int]`, but that type wouldn't allow for calling list-specific methods on the returned value, even though we're returning a list.
+
+Returning the most concrete type gives more flexibility to the caller to use that value in ways that the function/method author might not have considered.
+
+### Structural typing SHOULD NOT be used when possible.
+
+_Structural typing_ is a type system in which the structure of the type is what matters for type checking. Two values with the same structure will type-check as the same type, regardless of any type aliases being used (see the Python glossary for a stricter definition of [structural types](https://typing.python.org/en/latest/spec/glossary.html#term-structural)).
+
+Basic types like `str`, `list`, and `dict[int, str]` are examples of structural types in Python. Nothing about a `dict[int, str]` indicates what those keys or values represent, and any dictionary that takes integers as keys and has strings as values would type-check against that type.
+
+Type aliases are structural types; they DO NOT lead to nominal types. From the type checker's perspective, they are replaced with the type that they alias. They are only a convenience for not having to write out long structural types.
+
+Structural types should not be used when possible because they don't define the concepts that the types represent, reducing their usefulness.
+
+```python
+# Type alias for a quaternion
+# This doesn't define a new type, it just allows the function
+# definiton below to be shorter.
+Quaternion = Tuple[float, float, float float]
+
+def normalize_quaternion(quat: Quaternion):
+    # Is it this?
+    w, x, y, z = quat
+    # Or is it this?
+    x, y, z, w = quat
+    # do some normalization
+    return normalized_quaterion
+
+quat: Quaternion = (1.0, 0.0, 0.0, 0.0) # this is just of type `tuple`
+norm = normalize_quaternion(quat) # is this giving valid results?
+```
+
+Alternative ways to model a quaternion are using newtypes or dataclasses, both forms of _nominal types_. Which to choose would depend on whether additional functionality is needed, or for easier compatibility with third-party libraries.
+
+```python
+@dataclass
+class Quaternion:
+    w: float
+    x: float
+    y: float
+    z: float
+
+# Or
+QuaternionWXYZ = NewType("QuaternionWXYZ", Tuple[float, float, float, float])
```

Comment:
Do I understand it right that the main difference between this and `Quaternion = Tuple[float, float, float float]` is that we are giving it a more meaningful name? Or is there something else that the use of `NewType` adds?

Reply:
ah I think reading further, this is a nominal type now, right?

Reply:
Yes, it makes it a nominal type. I found it tricky to talk about structural typing without mentioning nominal typing and vice versa, so maybe I could expand this example a little bit to make it clearer.

The benefit of this is that you can't just use a raw tuple in places where a `QuaternionWXYZ` are needed and have it type check. The Python interpreter won't complain, because at runtime it's still just a tuple, but the type checker will mark it as an error.

In use, you'd have to do something like what happens in the UserID/AgentID example below where you wrap the raw tuple with a call to `QuaternionWXYZ()`.

Reply:
okay that makes sense. Maybe adding a simple "(more on NewType in the next section)" can defer questions of the curious reader :D